torch/_meta_registrations.py

```diff
@@ -143,3 +133,99 @@ def meta_repeat_interleave_Tensor(repeats, output_size=None):
             "cannot repeat_interleave a meta tensor without output_size"
         )
     return repeats.new_empty(output_size)
+
+@torch.library.impl(meta_lib, "index.Tensor")
+def meta_index_Tensor(self, indices):
+    # aten::index is the internal advanced indexing implementation
+    # checkIndexTensorTypes and expandTensors
+    result: List[Optional[Tensor]] = []
+    for i, index in enumerate(indices):
+        if index is not None:
+            check(
+                index.dtype in [torch.long, torch.int8, torch.bool],
+                lambda: "tensors used as indices must be long, byte or bool tensors"
+            )
+            if index.dtype in [torch.int8, torch.bool]:
+                nonzero = index.nonzero()
+                k = len(result)
+                check(
+                    k + index.ndim <= self.ndim,
+                    lambda: f"too many indices for tensor of dimension {self.ndim}",
+                    IndexError
+                )
+                for j in range(index.ndim):
+                    check(
+                        index[j] <= self.shape[k + j],
```

Comment:
In what case can index be less than the shape? I think these indices do not broadcast. This is implemented in `expandTensors` in `IndexingUtils.h` and they check for equality.

Reply:
lol I just copy pasted your code haha

Reply:
oh boi

Reply:
This modification to == seems to be wrong

Reply:
What's an example for this?

Reply:
Fwiw, in the implementation in TI it's an equality:
https://github.com/pytorch/pytorch/blob/519347df49fe50fb46f40544e22225be485efb06/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L478-L481

Reply:
I get this error:

```
======================================================================                                                        
ERROR: test_dispatch_meta___getitem___cuda_bfloat16 (__main__.TestMetaCUDA)                                                   
----------------------------------------------------------------------                                                        
Traceback (most recent call last):                                                                                            
  File "test/test_meta.py", line 325, in run_meta_crossref                                                                    
    meta_rs = func(*meta_args, **meta_kwargs)                                                                                 
  File "/raid/ezyang/pytorch-scratch2/torch/_ops.py", line 48, in __call__                                                    
    return self._op(*args, **kwargs or {})                                                                                    
  File "/raid/ezyang/pytorch-scratch2/torch/utils/_python_dispatch.py", line 71, in wrapped                                   
    return f(self, *args, **kwargs)                                                                                           
  File "/raid/ezyang/pytorch-scratch2/torch/utils/_python_dispatch.py", line 167, in __torch_dispatch__                       
    return func(*args, **kwargs)                                                                                              
  File "/raid/ezyang/pytorch-scratch2/torch/_ops.py", line 48, in __call__                                                    
    return self._op(*args, **kwargs or {})                                                                                    
  File "/raid/ezyang/pytorch-scratch2/torch/_meta_registrations.py", line 195, in meta_index_Tensor
    check(                                                                                                                    
  File "/raid/ezyang/pytorch-scratch2/torch/_prims/utils.py", line 1142, in check
    raise exc_type(s())                                                                                                       
IndexError: The shape of the mask torch.Size([5]) at index 0 does not match the shape of the indexed tensor torch.Size([5, 5, 
5]) at index 0 

```

```
  File "test/test_meta.py", line 336, in run_meta_crossref                                                                    
    raise RuntimeError(f"""\                                                                                                  
RuntimeError: failed to run: aten.index.Tensor(                                                                               
*(tensor(..., device='meta', size=(5, 5, 5), dtype=torch.bfloat16) stride=(25, 5, 1), [tensor([ True,  True, False,  True, Fal
se]) stride=(1,)]),                                                                                                           
**{}                                                                                                                          
)                                                                                                                             
   
```

Reply:
figured it out